components/region_cache_memory_engine/src/engine.rs

```diff
@@ -220,82 +243,227 @@ impl Iterable for RegionCacheMemoryEngine {
+        Ok(self.valid)
+    }
+
+    fn seek_for_prev_internal(&mut self, key: &[u8]) -> Result<bool> {
+        self.iter.seek_for_prev(key);
+        self.prev_internal();
+
+        Ok(self.valid)
+    }
+
+    fn prev_internal(&mut self) {
+        while self.iter.valid() {
+            let InternalKey { user_key, .. } = decode_key(self.iter.key());
+            self.saved_user_key.clear();
+            self.saved_user_key.extend_from_slice(user_key);
+
+            if user_key < self.lower_bound.as_slice() {
+                break;
+            }
+
+            if self.prefix_same_as_start {
+                // todo(SpadeA): support prefix seek
+                unimplemented!()
+            }
+
+            if !self.find_value_for_current_key() {
+                return;
+            }
+
+            self.find_user_key_before_saved();
+
+            if self.valid {
+                return;
+            }
+        }
+
+        // We have not found any key
+        self.valid = false;
+    }
+
+    // Used for backwards iteration.
+    // Looks at the entries with user key `saved_user_key` and finds the most
+    // up-to-date value for it. Sets `valid`` to true if the value is found and is
+    // ready to be presented to the user through value().
+    fn find_value_for_current_key(&mut self) -> bool {
+        assert!(self.iter.valid());
+        let mut last_key_entry_type = ValueType::Deletion;
+        while self.iter.valid() {
+            let InternalKey {
+                user_key,
+                sequence,
+                v_type,
+            } = decode_key(self.iter.key());
+
+            if !self.is_visible(sequence) || self.saved_user_key != user_key {
+                // no further version is visible or the user key changed
+                break;
+            }
+
+            last_key_entry_type = v_type;
+            match v_type {
+                ValueType::Value => {
+                    self.saved_value = Some(self.iter.value().clone());
+                }
+                ValueType::Deletion => {
+                    self.saved_value.take();
+                }
+            }
+
+            self.iter.prev();
+        }
+
+        self.valid = last_key_entry_type == ValueType::Value;
+        self.iter.valid()
+    }
+
+    // Move backwards until the key smaller than `saved_user_key`.
+    // Changes valid only if return value is false.
+    fn find_user_key_before_saved(&mut self) {
+        while self.iter.valid() {
+            let InternalKey { user_key, .. } = decode_key(self.iter.key());
+
+            if user_key < self.saved_user_key.as_slice() {
+                return;
+            }
+
+            self.iter.prev();
+        }
+    }
+}
+
 impl Iterator for RegionCacheIterator {
     fn key(&self) -> &[u8] {
         assert!(self.valid);
-        self.iter.key().as_slice()
+        &self.saved_user_key
     }
 
     fn value(&self) -> &[u8] {
         assert!(self.valid);
+        if self.saved_value.is_some() {
```

Comment:
nit: is it possible to re-write this as:
```rust
self.saved_value().map_or_else(|| self.iter_value().as_slice(), |v| v.as_slice()
```
or

```rust
if let Some(saved_value) = self.saved_value.as_ref() {
    saved_value.as_slice()
} else {
  self.iter.value().as_slice()
}
```

Reply:
Good suggestions and modified.